_posts/2020-10-24-spring-boot-specification.md

```diff
@@ -0,0 +1,372 @@
+```
+
+Finally, to execute our query we can simply call:
+
+```java
+List<Product> products = productRepository.findAll(namelike("reflectoring"));
+```
+
+We can also take benefit of `findAll` functions overloaded with `Pageable` and `Sort` in case we are expecting a large number of records in the result or
+want records in sorted order.
+
+Specification interface also has some public static helper methods - `and`, `or`, `where` - that allows us to combine 
+multiple specifications. Also, a `not` method which allows us to do negation of a specification.
+
+Let's look at an example:
+
+```java
+public List<Product> getPremiumProducts(List<Category> categories) {
+  return productRepository.findAll(where(belongsToCategory(categories))
+                                                    .and(isPremium()));
+}
+
+private Specification<Product> belongsToCategory(List<Category> categories){
+  return (root, query, criteriaBuilder)-> 
+            criteriaBuilder.in(root.get(Product_.CATEGORY)).value(categories);
+}
+
+private Specification<Product> isPremium() {
+  return (root, query, criteriaBuilder) ->
+          criteriaBuilder.and(
+                  criteriaBuilder.equal(
+                          root.get(Product_.MANUFACTURING_PLACE).get(Address_.STATE),
+                          STATE.CALIFORNIA),
+                  criteriaBuilder.greaterThanOrEqualTo(
+                          root.get(Product_.PRICE), PREMIUM_PRICE));
+}
+```
+
+Here, we have combined `belongsToCategory` and `isPremium` specifications into one using `where` and `and` helper functions. 
+This also reads really nice right? Also, notice how `isPrimium` is giving more meaning to the query.
+
+Currently `isPremium` is combining two predicates, but if we want we can create separate Specification for each of those 
+and combine again with Specification `and`. For now, we will keep it as is as the predicates used in `isPremium` are very 
+specific to that query, and if in the future we need to use them in other queries too then we can always split them up without 
+impacting the clients of `isPremium` function.
+
+## Dynamic Queries With Specification
+
+Let's say we want to allow our users to filter the products based on a number of properties such as categories, price, color, 
+etc. Here we do not know beforehand what combination of properties our user is going to use, to filter the product. One way 
+to handle this is to write query methods for all possible combinations but that would require writing a lot of query methods. And that number would further increase as we introduce new fields.
+
+A better solution is to generate queries dynamically using the specification. Let's see how.
+
+First, let's create an input object to take filters from the clients:
+
+```java
+public class QueryInput {
+  private String field;
+  private QueryOperator operator;
+  private String value;
+  private List<String> values;//Used in case of IN operator
+  private boolean isOptional;
```

Comment:
The `isOptional` field is a bit strange here. The `and` or `or` operators should be a level above the filter, because we want to combine multiple filters with `and` or `or`.

I see that you're setting `isOptional` to `false` anyways in the example, so I would suggest to remove this field from the example code and assume that users want to combine filters with `and`.